rfcs/20190904-tfx-generic-container-based-component.md

```diff
@@ -0,0 +1,863 @@
+hood, as a specialized subclass of it. Such helper shall eventually converge to
+the other way of implementing a
+[custom component](https://github.com/tensorflow/tfx/tree/80a80ffb0f833cf02344ac2898017cf1b73e166d/tfx/examples/custom_components)
+for TFX via a custom `Executor` class written in Python, and package it in a
+container image to release.
+
+Detailed RFC particularly on this point will follow.
+
+### Component Arhcetypes
+
+As of `tfx` 0.14, there are
+[10 known artifact types](https://github.com/tensorflow/tfx/blob/767968262233bb8d37837e8de9e11d1065b07c7b/tfx/types/standard_artifacts.py)
+defined and used.
+
+*   ExternalArtifact
+*   Examples
+*   Schema
+*   ExampleStatistics
+*   ExampleAnomalies
+*   TransformGraph
+*   Model
+*   ModelEvalulation
+*   ModelBlessing
+*   PushedModel
+
+Based on the above known artifact types, TFX defines the following
+[9 component archetypes](https://github.com/tensorflow/tfx/tree/767968262233bb8d37837e8de9e11d1065b07c7b/tfx/components).
+
+| **Component**    | **Inputs**                | **Outputs**              |
+| :--------------- | :------------------------ | :----------------------- |
+| ExampleGen       | ExternalPath (optional)   | Examples                 |
+| StatisticsGen    | Examples                  | ExampleStatistics        |
+| SchemaGen        | ExampleStatistics         | Schema                   |
+| ExampleValidator | ExampleStatistics, Schema | ExampleAnomalies         |
+| Transform        | Examples (raw), Schema    | TransformGraph, Examples |
+:                  :                           : (transformed)            :
+| Trainer          | Examples, TransformGraph  | Model                    |
+:                  : (optional), Schema        :                          :
+| Evaluator        | Examples, Model           | ModelEvaluation          |
+| ModelValidator   | Examples, Model           | ModelBlessing            |
+| Pusher           | Model, ModelBlessing      | PushedModel              |
+
+The proposed generic container-based component will enable scenarios where, so
+long as the wrapped component adheres to one of the above the input- and output-
+archetypes, it enables drop-in replacement, while retaining interactions with
+the rest of the components in the pipeline, and also the pipeline not having to
+know the actual business logic inside the container application.
+
+### Specification of Artifacts
+
+As of tfx 0.14, the schema (list of properties) of metadata for each artifact
+type is defined implicitly when it is created and used
+([example](https://github.com/tensorflow/tfx/blob/767968262233bb8d37837e8de9e11d1065b07c7b/tfx/components/transform/component.py#L99)).
+
+In order for the proposed generic component-based component to utilize artifacts
+and its metadata in a standardized way, such metadata schema definition needs to
+be made explicit, possibly as a Python class. In other words, unless the known
```

Comment:
`possibly as a Python class`, would you please provide some example or pointers here?

Reply:
Added pointer to current Artifact base class implementation, which prescribes the default set of properties.